Add bevy_light::SunDisk

defuz opened the pull request:
# Objective

* Fixes **#20425** (GI probes capture the sun disk while a `DirectionalLight` also lights the scene, so we need a way to skip the disk during probe baking).
* ~~**Add `sun_disk_intensity: f32` to `AtmosphereSettings`** so you can smoothly hide, dim, or over‑brighten the sun disk without affecting the visible illumination of scene objects, while still allowing the sun’s radiance to scatter into the atmosphere even when the disk itself is hidden.~~
*  Add a bevy_light::SunDisk component that acts as an add‑on to DirectionalLight. https://github.com/bevyengine/bevy/pull/20434#issuecomment-3190804227

## Testing

* Ran the official `atmosphere` example on a MacBook M1 (see showcase below).
* Toggled bloom on/off and tried intensities 0, 0.01, 1, and 10 — the sky looks correct and no warnings appear.
* Noticed a small dark “hole” in the center of the sun disk under some settings; this artifact was present before these changes (often hidden by bloom) and probably deserves further investigation. (FIXED IN THIS PR)

I’m new to WGSL/shader code — please point out anything silly! 🙂

---

## Showcase

* Scene brightness changes with intensity only because bloom amplifies the bright disk; sky illumination stays unchanged.
* With bloom disabled, the disk is the only thing that changes — the rest of the scene looks identical.
* Raising intensity above 1.0 doesn’t affect the scene without bloom (the disk is already 100 % white) but boosts the bloom halo when bloom is enabled.
* Lowering intensity reduces the bloom halo; drop it far enough and the disk blends into the sky, which is physically plausible for very dim suns.

<details>
  <summary><strong>1.0</strong> (bloom on) — normal</summary>
  <img width="1680" height="1050" alt="default 1 0" src="https://github.com/user-attachments/assets/89144e52-1159-42c6-9325-f80f986349ec" />
</details>

<details>
  <summary><strong>0.0</strong> (bloom on) — no disk</summary>
  <img width="1680" height="1050" alt="no sun 0 0" src="https://github.com/user-attachments/assets/2b3842ae-26ed-45bf-9266-20189bc487a7" />
</details>

<details>
  <summary><strong>0.01</strong> (bloom on) — low brightness, tiny bloom effect</summary>
  <img width="1680" height="1050" alt="0 21" src="https://github.com/user-attachments/assets/8c3943df-74b6-4422-8c9b-061e56ee33f8" />
</details>

<details>
  <summary><strong>0.00001</strong> (bloom on) — disk blends into the sky</summary>
   <img width="1680" height="1050" alt="0 00001" src="https://github.com/user-attachments/assets/e1228131-41b2-4f30-b50d-ebceebc8cabb" />
</details>

<details>
  <summary><strong>10.0</strong> (bloom on) — blinding glare</summary>
  <img width="1680" height="1050" alt="10" src="https://github.com/user-attachments/assets/3527c6d6-27ec-4e6f-9e9f-564a6b246bba" />
</details>

<details>
  <summary><strong>1.0</strong> (bloom off) — normal sun with no bloom</summary>
  <img width="1680" height="1050" alt="1 no bloom" src="https://github.com/user-attachments/assets/b62ce660-cb65-46b1-b211-a5d330b08719" />
</details>

<details>
  <summary><strong>10.0</strong> (bloom off) — indistinguishable from 1.0 when bloom is off</summary>
  <img width="1680" height="1050" alt="10 no bloom" src="https://github.com/user-attachments/assets/17ea20dd-3de7-4bb9-82a1-16f7df9a9994" />
</details>

Quick question for reviewers: (not part of this PR) I’ve already experimented with parameterizing the **sun angular size** and it works well. It could serve as a cool artistic control or let you match skies for planets that aren’t Earth. Would folks be interested in a follow‑up PR?

<details>
  <summary>4× sun size</summary>
  <img width="1680" height="1050" alt="4x sun" src="https://github.com/user-attachments/assets/5c39ef98-0376-48b7-b845-bfad6d87c09b" />
</details>

alice-i-cecile commented:
> Quick question for reviewers: (not part of this PR) I’ve already experimented with parameterizing the sun angular size and it works well. It could serve as a cool artistic control or let you match skies for planets that aren’t Earth. Would folks be interested in a follow‑up PR?

I'm very interested in this! Seems extremely useful artistically and very sensible to expose.

mate-h commented:
Hey everyone, this is a very neat! Thank you for putting together this PR. I do have a pending change to include the angular radius in directional light, as well as a new SunLight component. I could tac on this change to the SunLight component itself rather than the atmosphere settings. So there might be some conflicts. I will leave a more detailed response with my specific suggestions for the user facing API later as I have thought extensively about this problem based on all the feedback received so far. Also regarding integrating the atmosphere with probes, another problem area that I have addressed in my fork of Bevy, still needs to be reviewed and merged to main.

mate-h commented:
What is your current approach with GI probe cameras? I am also proposing to add another component called atmosphere environment map light that does that part. It is meant to upsample the sky view LUT (or do a raymarch) into a target cubemap texture, in a separated shader without the sun disk. Then it is plugged into the (recently merged) environment map filtering pipeline to get cubemap mip chain at different roughness levels. I could still see the need for a separate decoupled intensity setting for the sun disk itself. But it would only affect the composited sky in the main sky rendering pass, not the environment light. There the disk is deliberately excluded, and instead we sample the transmittance LUT in the pbr pipeline to tint the directional lights orange at sunset. In its present state Bevys atmosphere is only a composited effect but the changes I am proposing would fully integrate it with the PBR pipeline and environment reflections. Feel free to reach out on Discord (user mate_h) to discuss the approach and the details.

mate-h commented:
There is a newer , working version https://github.com/mate-h/bevy/pull/10
Rebased onto more recent main. 

But I need to work on splitting it up a bit more.

JMS55 commented:
Angular sun disk size for directional lights is also something I'm interested in. The code for it is already in Solari, it's just not exposed to users :)

JMS55 commented on crates/bevy_pbr/src/atmosphere/mod.rs:
```diff
@@ -378,6 +378,13 @@ pub struct AtmosphereSettings {
+    ///
+    /// `0.0` disables the sun disk entirely while still
+    /// allowing the sun's radiance to scatter into the atmosphere,
+    /// and `1.0` renders sun disc at its normal intensity.
```
```suggestion
    /// and `1.0` renders the sun disk at its normal intensity.
```

JMS55 reviewed (approved):
Approved once the wording change I proposed is merged

ecoskey commented:
I like the idea of having a stylistic adjustment for the sun-disc intensity, but I think I'd prefer something like @mate-h's proposed `SunLight` component. It feels easier to teach and reason about, even if a bit harder to implement :)

Also (and this is more editorial) maybe the intensity override should be an `Option`? I think in most cases we should nudge people towards matching the values.

defuz commented:
After discussing with @mate-h on Discord, I reworked the approach to expose sun‑disk controls. Instead of an "atmosphere setting" for disk intensity, this PR now adds a `bevy_light::SunLight` component that acts as an add‑on to `DirectionalLight`. 

The intent is to keep this forward‑compatible with mate-h/bevy/pull/10.

### 1) **(BREAKING)** Physically correct default angular diameter

We now define the default **angular diameter** of the solar disk as **`0.00930842` radians** (≈ 32 arcminutes), which matches the mean apparent size of the Sun at 1 AU.

The previous constant in the shader was **`0.0174533`** (≈ 1 degree), roughly **2× larger** than the real Sun. With the corrected default, the rendered disk appears \~**2× smaller** (as it should).

### 2)  **(BREAKING)** Correct sun‑disk solid angle

In the small‑angle approximation, solid angle should be computed from the **half‑angle**:

$\Omega \approx \pi \times radius^2 = \pi \times diameter^2/4$

**Before**

```wgsl
let sun_solid_angle = (SUN_ANGULAR_SIZE * SUN_ANGULAR_SIZE) * 4.0 * FRAC_PI;
```

**After**

```wgsl
let sun_solid_angle = (sun_angular_size * sun_angular_size) * 0.25 * PI;
```

This makes the disk radiance **physically correct**. Numerically, the previous expression overestimated Ω by \~**1.62×**, so the disk appeared \~**38% dimmer** than it should. With the fix, the disk becomes \~**62–64% brighter** (depending on rounding), which matches the expected correction.

I kept the small‑angle approximation for simplicity. If preferred, we can switch to the exact expression `Ω = 2π(1 − cos(α))`.


### 3) Clamp `acos` input to avoid NaNs at the disk center

**Before**

```wgsl
let angle_to_sun = fast_acos(neg_LdotV);
```

**After**

```wgsl
let angle_to_sun = fast_acos(clamp(neg_LdotV, -1.0, 1.0));
```

This prevents NaN propagation from floating‑point overshoot (e.g., `1.0000001`) when looking directly at the Sun, which previously caused a **hole at the disk center** on unlucky view camera angles/drivers.

<img width="216" height="170" alt="Screenshot 2025-08-15 at 04 59 59" src="https://github.com/user-attachments/assets/cdf58252-8844-44ba-93cf-76e54fc198d4" />

### 4) Unbiased, energy‑preserving limb anti‑aliasing

**Before** (one‑sided ramp, reversed edges)

```wgsl
let factor = smoothstep(0.0, -pixel_size * ROOT_2, angle_to_sun - SUN_ANGULAR_SIZE * 0.5);
```

**After** (centered ramp around the true edge)

```wgsl
let w = max(0.5 * fwidth(angle_to_sun), 1e-6);
let factor = 1.0 - smoothstep(sun_angular_size * 0.5 - w,
                              sun_angular_size * 0.5 + w,
                              angle_to_sun);
```

Centering the transition band on the boundary removes the systematic radius shrink and keeps total disk power essentially constant across resolution/FOV. Without this, the sun is slightly dimmer then should be, especially on low resolutions.

---

## Resulting disk radiance (what we render)

With the fixes in place, the emitted (scene‑linear, HDR) radiance we add for the solar disk is:

```text
sun_disk.rgb = directional_light.color * directional_light.illuminance
               / (π * sun_light.angular_size^2 / 4)
               * sun_light.intensity
```

* `sun_light.angular_size` is the **diameter** (radians)
* `sun_light.intensity` is a **unitless artistic multiplier**.

This makes the disk’s photometric Y proportional to `illuminance / Ω`, which is the physically correct relationship (lux → cd/m² via division by solid angle).


## Questions for reviewers / maintainers
### 1) Should `bevy_light::SunLight` be re‑exported from the prelude?

### 2)  Color normalization semantics. 

Bevy’s docs aren’t explicit about the physical meaning of `DirectionalLight::color` and `DirectionalLight::illuminance`. If `color` is intended purely as a **tint**, it should be Y‑normalized so changing hue doesn’t change illuminance:

   ```rust
   // linear‑sRGB tint
   let tint = Vec3::new(r, g, b);
   let y = 0.2126*tint.x + 0.7152*tint.y + 0.0722*tint.z;
   let colorY = if y > 0.0 { tint / y } else { Vec3::splat(1.0) };
   // Store color premultiplied by illuminance (lux)
   light.color = Vec4::new(colorY.x, colorY.y, colorY.z, 1.0) * light.illuminance; // Y == illuminance
   ```

Without this, `illuminance` effectively becomes a **unitless scale** and the actual photometric output varies with tint (all non‑white colors change brightness). I did **not** change this behavior in this PR because other parts of the pipeline may rely on the current semantics, but it would be helpful to clarify the intended contract in documentation.

### 3) **Default behavior when `SunLight` is absent.**

   In this PR, when a `DirectionalLight` has no `SunLight`, we currently render a default disk (using Earth‑like values). This is  backward compatibile, but can be *surprising* behavior, especially if a scene has **multiple** directional lights. An alternative would be **absence = disabled** (i.e., don’t render a disk unless the component is present). I’m happy to switch to that if preferred.

### 4)  **`SunLight::intensity` vs. user‑defined absolute brightness/tint.**

   Currently `intensity` is a unitless multiplier applied to the physically derived disk radiance, which works well for “take the correct Sun and scale it by X%”. Some users may prefer a direct control that’s independent of `::illuminance`, or `angular_size`, or the light tint. Possible API variants:

1. `SunLight { intensity: f32, .. }` *(current behavior; unitless)*
  Scales the physically derived disk radiance. The disk’s final color depends on `DirectionalLight::illuminance`, `DirectionalLight::color`, and `SunLight::angular_size`. `intensity` simply multiplies the result.

2. `SunLight { disk_luminance: Option<f32>, .. }` *(absolute nits)*
  If `Some(L)`, render the disk at exact luminance (cd/m²). This makes the disk independent of `DirectionalLight::illuminance` and `SunLight::angular_size`. The **hue** is still taken from `DirectionalLight::color` (tint), so the Sun’s color tracks the light unless we introduce a separate disk tint.

3. `SunLight { disk_illuminance: Option<f32>, .. }` *(absolute lux)*
  If `Some(E)`, treat this as the Sun’s direct‑normal illuminance and compute `L = E / Ω` internally. The disk becomes independent of `DirectionalLight::illuminance`, but remains dependent on `SunLight::angular_size`, and still uses `DirectionalLight::color` for tint.

4. `SunLight { disk_color: Option<Vec3>, .. }` *(full override)*
  If `Some(rgb)`, write this scene‑linear RGB directly for the disk. This bypasses `illuminance`, `angular_size`, and `color`, and **is the only option that lets you change the Sun’s tint without affecting scene lighting (e.g., “more yellowish Sun” without warming the world).**

## What can be further improved (not part of this PR currently)

1. We can eliminate `acos` completely for potentially better performance:

  ```wgsl
  let cosθ = clamp(dot((*light).direction_to_light, ray_dir_ws), -1.0, 1.0);
  let cosα = cos(0.5 * (*light).angular_size);
  let w    = max(fwidth(cosθ), 1e-6);
  // inside the disk means θ <= α  <=>  cosθ >= cosα
  let factor = smoothstep(cosα - w, cosα + w, cosθ);
  ```

2. Exact calculation of disk solar angle `Ω = 2π(1 − cos(α))` (might be important for HUGE suns).

3. We currently assume the same brightness across the whole disk (at the center and at the edges), which is not physically accurate. In reality, the limb is darker than the center. We could go further with sun‑disk rendering, but it’s unlikely to impact 99% of users.

cc @alice-i-cecile @mate-h @JMS55 @ecoskey

JMS55 commented on crates/bevy_light/src/directional_light.rs:
```diff
@@ -240,3 +240,38 @@ pub fn update_directional_light_frusta(
     }
 }
+
+/// This component marks a directional light entity for changing the size and intensity of the sun disk.
```
Please rustdoc-link to DirectionalLight.

JMS55 commented:
Do you mind also hooking this up to Solari? Just needs an adjustment to the Rust code to query for `Option<&SunLight>` and pass the size to the existing code:

https://github.com/bevyengine/bevy/blob/a76e0a20d9ea44f9f042e15a8624024f41e1f5dd/crates/bevy_solari/src/scene/binder.rs#L410

JMS55 commented:
Fyi you can probably delete SUN_ANGULAR_DIAMETER_RADIANS now. Not sure why CI didn't warn that it's now unused 🤔.

mate-h reviewed (commented):
Great work on the new approach! looks solid with no blockers from my side.
- Shadows vs sun disk size: Keep these independent in the raster pipeline for finer control.
- Preserve physically-based default (intensity = 1.0 → physically derived values). Avoid extra `SunLight` APIs (e.g., separate color). For stylized/colored/extraplanetary disks, recommend a custom skybox or compute-to-cubemap; the atmosphere can composite over via dual-source blending.
- Disk intensity falloff: Physically motivated but likely low visual impact—defer as a later experiment to avoid API bloat.
- I sppreciated the forward-compatibility, correct fwidth AA, and center NaN fix. Feels robust and production-ready. Keep defaults to avoid breakage while offering customization; don’t over-expose knobs.
- Directional light: For Solari, tying shadow blur to disk size makes sense. Consider adding angular size directly to `DirectionalLight` or introducing `DirectionalDiskLight` for generic infinite disk lights. For the Moon, prefer cubemap/skybox.
- Please rename the component to `SunDisk` (it’s not a new light type). I’ll approve after the rename.
- Sun disk rendering is niche; now that it’s decoupled from the atmosphere, consider generic use cases. Ensure users can disable the built-in disk and supply their own; sensible defaults should cover most needs.

mate-h commented on crates/bevy_pbr/src/atmosphere/functions.wgsl:
```diff
@@ -285,11 +282,15 @@ fn sample_sun_radiance(ray_dir_ws: vec3<f32>) -> vec3<f32> {
+        let sun_intensity = (*light).intensity;
+        if sun_angular_size > 0.0 && sun_intensity > 0.0 {
+            let factor = 1 - smoothstep(sun_angular_size * 0.5 - w, sun_angular_size * 0.5 + w, angle_to_sun);
+            let sun_solid_angle = (sun_angular_size * sun_angular_size) * 0.25 * PI;
```
This is the same expression as before, even though the PR description suggests otherwise.
FRAC_PI = 1 / PI

0.25 = 1 / 4

defuz replied:
It's precisely reciprocal, no? 4/pi vs pi/4. :)

mate-h commented on crates/bevy_light/src/directional_light.rs:
```diff
@@ -240,3 +240,38 @@ pub fn update_directional_light_frusta(
+}
+
+impl SunLight {
+    pub const SUN: SunLight = SunLight {
```
typing out `SunLight:SUN` or `SunDisk:SUN` feels a bit awkward, but I don't have a better idea. maybe `SunDisk:EARTH` ? since this is exactly 1 AU away? I am adding martian atmosphere eventually so there can be a `SunDisk:MARS` too that has different values.

JMS55 replied:
Maybe `SunDisk::VIEWED_FROM_EARTH`

JMS55 commented:
> You are right, but the same argument is also applicable to SunLight::angular_size. If I understand correctly, SunLight::angular_size affects GpuDirectionalLight in a way that changes how soft shadows look with Solari.

Yes, but Solari doesn't render a sun disk, so the SunDisk::intensity field shouldn't matter, right? Whereas angular_size _does_ affect something that Solari uses.

> But bevy::pbr renders soft shadows relying on DirectionalLight::soft_shadow_size: Option<angular_size> and is independent of SunLight::angular_size entirely, which is a bit awkward, but it represents a separation between the shadow and sky rendering implementations, while Solari rendering implementation is generally more “physically uniform”, which means users have fewer options to tweak things separately.

Yes, it's a little awkward. For now just leave it. PCSS (which is what soft_shadow_size is used for) were never really working anyways.

> As result, in the current implementation, Solari relies on SunLight::angular_size for sun/world rendering, and ignores SunLight::intensity for sun/world rendering. And the classic render pass relies on both SunLight::angular_size/intensity for sun rendering, and ignores both for world rendering. This looks like a counterintuitive blend of two paradigms in usage of same entity.

Solari doesn't render the sun disk, so intensity doesn't matter.

The standard renderer doesn't render the sun disk, _or_ have physically accurate directional light shadows, so both angular_size and intensity don't matter.

JMS55 commented on crates/bevy_pbr/src/render/light.rs:
```diff
@@ -103,6 +104,8 @@ pub struct ExtractedDirectionalLight {
     /// True if this light is using two-phase occlusion culling.
     pub occlusion_culling: bool,
+    pub angular_size: f32,
+    pub intensity: f32,
```
I'm fine with leaving it as angular_size, but definitely rename it to sun_disk_intensity.

defuz commented:
Hi folks! Thanks a lot for all your comments. They were extremely helpful to me, especially in narrowing down the approach and aligning it with Bevy’s vision. Also very insightful!

---

* Renamed `SunLight` to `SunDisk`.
* Renamed the constant `::SUN` to `::EARTH`.
* **Not discussed previously:** Added a `SunDisk::OFF` constant (`angular_size = 0`, `intensity = 0`). I think it fits well with Bevy’s approach and provides a very intuitive answer to the question “How do I disable sun-disk rendering for my directional light?” → “Just add `SunDisk::OFF` to it.”
* Rename fields in `ExtractedDirectionalLight` to reduce ambiguity.

For now, I don’t see anything else that needs to be fixed in the PR before merging. If I’m missing anything, please let me know!

---

@mate-h Your comments look 100% reasonable to me; I have nothing to add.

@JMS55

> Solari doesn't render the sun disk, so intensity doesn't matter.

You’re right. I misunderstood this initially. Now everything looks much more reasonable to me.

JMS55 commented on crates/bevy_light/src/directional_light.rs:
```diff
@@ -240,3 +240,55 @@ pub fn update_directional_light_frusta(
+}
+
+impl SunDisk {
+    /// Parameters of the Sun disk as viewed from Earth.
```
```suggestion
    /// Parameters of the sun disk as viewed from Earth.
```

JMS55 commented on crates/bevy_light/src/directional_light.rs:
```diff
@@ -240,3 +240,55 @@ pub fn update_directional_light_frusta(
     }
 }
+
+/// This component marks a [`DirectionalLight`] entity for changing the size and intensity of the sun disk.
```
This sentence is a little weird sounding to me, but I don't know what would be better.

JMS55 commented:
I don't have the knowledge to review the atmosphere changes, but the rest is almost ready! Just some minor doc suggestions.
